src/enumo/ruleset.rs

```diff
@@ -432,18 +436,28 @@ impl<L: SynthLanguage> Ruleset<L> {
         }
     }
 
-    pub fn minimize(&mut self, prior: Ruleset<L>, scheduler: Scheduler) -> Self {
+    pub fn minimize(&mut self, prior: Ruleset<L>, scheduler: Scheduler) -> (Self, Self) {
+        let before = self.len();
+        let mut invalid: Ruleset<L> = Default::default();
         let mut chosen = prior.clone();
         let step_size = 1;
         while !self.is_empty() {
-            let selected = self.select(step_size);
+            println!(
+                "Shrinking {}/{} candidates. Kept {} so far.",
+                self.len(),
+                before,
+                chosen.len()
+            );
+            let selected = self.select(step_size, &mut invalid);
+            print!("Selected: ");
```

Comment:
nit: remove printlns if you don't need them anymore
(fine to do as followup if you want to get this merged soon!)